src/main.rs

```diff
@@ -0,0 +1,600 @@
+                                panic!("Only unsecure http:// proxies are supported");
+                            }
+                            url
+                        },
+                    Err(err) => panic!("Invalid proxy url: {}, only urls on the format \"http://host:port\" are allowed", err)
+                    }
+                },
+            ),
+            ap_port: matches
+                .opt_str("ap-port")
+                .map(|port| port.parse::<u16>().expect("Invalid port")),
+        }
+    };
+
+    let player_config = {
+        let bitrate = matches
+            .opt_str("b")
+            .as_ref()
+            .map(|bitrate| Bitrate::from_str(bitrate).expect("Invalid bitrate"))
+            .unwrap_or(Bitrate::default());
+        let gain_type = matches
+            .opt_str("normalisation-gain-type")
+            .as_ref()
+            .map(|gain_type| {
+                NormalisationType::from_str(gain_type).expect("Invalid normalisation type")
+            })
+            .unwrap_or(NormalisationType::default());
+        PlayerConfig {
+            bitrate: bitrate,
+            gapless: !matches.opt_present("disable-gapless"),
+            normalisation: matches.opt_present("enable-volume-normalisation"),
+            normalisation_type: gain_type,
+            normalisation_pregain: matches
+                .opt_str("normalisation-pregain")
+                .map(|pregain| pregain.parse::<f32>().expect("Invalid pregain float value"))
+                .unwrap_or(PlayerConfig::default().normalisation_pregain),
+        }
+    };
+
+    let connect_config = {
+        let device_type = matches
+            .opt_str("device-type")
+            .as_ref()
+            .map(|device_type| DeviceType::from_str(device_type).expect("Invalid device type"))
+            .unwrap_or(DeviceType::default());
+
+        let volume_ctrl = matches
+            .opt_str("volume-ctrl")
+            .as_ref()
+            .map(|volume_ctrl| VolumeCtrl::from_str(volume_ctrl).expect("Invalid volume ctrl type"))
+            .unwrap_or(VolumeCtrl::default());
+
+        ConnectConfig {
+            name: name,
+            device_type: device_type,
+            volume: initial_volume,
+            volume_ctrl: volume_ctrl,
+            autoplay: matches.opt_present("autoplay"),
+        }
+    };
+
+    let enable_discovery = !matches.opt_present("disable-discovery");
+
+    Setup {
+        backend: backend,
+        cache: cache,
+        session_config: session_config,
+        player_config: player_config,
+        connect_config: connect_config,
+        credentials: credentials,
+        device: device,
+        enable_discovery: enable_discovery,
+        zeroconf_port: zeroconf_port,
+        mixer: mixer,
+        mixer_config: mixer_config,
+        player_event_program: matches.opt_str("onevent"),
+        emit_sink_events: matches.opt_present("emit-sink-events"),
+    }
+}
+
+#[tokio::main]
```

Comment:
I'm not sure what we were doing before but maybe we should change this? This is going to start one OS thread per logical core for tokio. On my machine that'd be 16 threads. Seems like librespot probably doesn't need those kinds of resources.

Reply:
So what do you think is a reasonable number of threads? I think with tokio-core it was only single-threaded.

Reply:
Single threaded sounds good to me, you can do this with the `current_thread` builder. I'm not sure if there's a macro associated with this.

Reply:
The thread count can be configured using `#[tokio::main(flavor = "multi_thread", worker_threads = 10)]`. To get the single thread behaviour back, `#[tokio::main(flavor = "current_thread")]` can be used. Would it be worth adding a CLI argument such as `--multithread XX` where `XX` is the thread count desired by the user, otherwise defaulting to single thread operation? We'd probably need to check user input didn't exceed the number of system cores if we opted to do this.

Reply:
Aren't we anyway network bound? Not sure if having more threads is going to help much. Although would be great to have some numbers to compare. :-)

Reply:
Whilst I'm not too sure on the specifics, I'm wondering if machines such as the RPi might benefit from multithreading. Do they even have multiple cores or are they single?

Reply:
Usually you spawn the multithreaded environment when you need to handle a lot of traffic. librespot is not meant to be handling like 10's of thousands of requests per second right? I think of the single-threaded tokio scenario like a node runtime, you get an event loop and concurrency, but you don't need the throughput that requires all the cores on your machine.

Reply:
making it single threaded is easy enough, `#[tokio::main(flavor = "current_thread")]` should do it as mentioned above.

Reply:
Just to add, making it configurable seems like it would be fine too, but something that may best be saved for a future PR? I've written a few tokio services where you pass a cli or config param that says how many threads the runtime should use.

Reply:
probably a good idea, in case there are any modifications required elsewhere to accommodate it.

Reply:
Many people try hard to use all cores in the best way, and encounter strange bugs while trying this. We don't have concurrency bugs, but still we're discussing whether to make it single-threaded. 😀

Reply:
I'm not sure I follow, the type of runtime you want to use will be dictated by your requirements. Nothing about the project README indicates that we want a high-throughput server capable of handling tens of thousands of req/s, so I think having the default configuration such that the footprint is minimal is good for now.

Reply:
I was mainly joking. It would be interesting to see the differences between different numbers of threads, but I agree, 16 threads are not necessary.

Reply:
Maybe you would like to create a PR? (the commit history of this branch looks too monotonous anyway)

Reply:
Sorry my comment went to the wrong place, sarcasm is sometimes hard over the internet :smile:  Do you want to just update this PR to use `[tokio::main(flavor = "current_thread")]` I think that would be sufficient if @sashahilton00 agrees